### DISCUSSION TITLE
<Image> performance with next start

### DISCUSSION THREAD

**Discussion by schlosser** [vercel/next.js#18481] [Help] [answered]
Hi there! 

I'm upgrading my project to v10 on a branch, and using `next start` to deploy my app. I noticed that the images on my site are somewhat slow to load, even after the first time I load them at a screen size. 

When using `next start` with local images (in `/public`), are the resized images cached anywhere? Or does compression + resizing happen on every requests?

I noticed that [the RFC](https://github.com/vercel/next.js/discussions/16832) mentions

> Though the Image component is designed with CDNs in mind, it will also work with locally hosted images. The application developer simply needs to add an image compression/resizing step into their build process, and write a custom loader that generates URLs pointing to the output paths used by that build process.

Would anyone be able to share steps to do this, if it's necessary for `next start`? Thanks in advance!

**Comment by synoptase** [comment]
I've noticed that too. There's no `Cache-Control` on the response header. This means that serving images from the embed express server is not very efficient. I'm wondering if one can use the [headers feature](https://nextjs.org/docs/api-reference/next.config.js/headers) 🤔 
Thank you for pointing out the RFC. I'd be interested with any solution you come up with and will definitely check out custom loaders when i get some time.

**edit**:
there's an internal cache that is managed by the image-optimizer. [It defaults to 60](https://github.com/vercel/next.js/blob/canary/packages/next/next-server/server/image-optimizer.ts#L213). Haven't found out yet if the [getRequestHandler ](https://github.com/vercel/next.js/blob/canary/packages/next/next-server/server/image-optimizer.ts#L208) that is fetching the image sets a `Cache-Control` header. 

**edit 2**:
@schlosser , to finally answer your question:
- there is an internal caching mechanism. By default it uses a 60s delay for locally hosted images, otherwise it tries to parse the `Cache-Control` header of the supplied url. In both of these cases, it will locally store the image and use the `max-age` value of `Cache-Control` as the expiry time.
- there isn't any `Cache-Control` header sent from the the `/_next/images` endpoint, therefore browser cache isn't leveraged :( Why is that: 🤷‍♀️

**Reply by schlosser** [reply]
Yeah, this makes `<Image>` pretty unusable for me unfortunately..

**Comment by synoptase** [comment, marked as answer]
Heads up @schlosser : https://github.com/vercel/next.js/pull/18399#issuecomment-724235637

**Reply by schlosser** [reply]
aha! Thank you for pointing this out; this fixes my issue, as I can address the upstream headers (or lack thereof) that was likely causing this. Thank you!

**Reply by schlosser** [reply]
and when that is merged I'll do even better :)

**Reply by schlosser** [reply]
Hmm... I'm thinking through this a bit more and reading through the logic for [`getMaxAge`](https://github.com/vercel/next.js/blob/50400f7eebfa1f2ff11b2f662d79e4e7ea54de7b/packages/next/next-server/server/image-optimizer.ts#L348-L362), and now I'm a bit confused. 

If I have code like:

```jsx
<Image
  src="/img/usa-map.png"
  alt="Map of the United States"
  width={674}
  height={426}
/>
```

And a `/public/img` folder with my `usa-map.png` in it, what would the max-age be? 

It looks like `isAbsolute` would be [set to `false`](https://github.com/vercel/next.js/blob/50400f7eebfa1f2ff11b2f662d79e4e7ea54de7b/packages/next/next-server/server/image-optimizer.ts#L68-L70), and then a mock response with [mock headers is created](https://github.com/vercel/next.js/blob/50400f7eebfa1f2ff11b2f662d79e4e7ea54de7b/packages/next/next-server/server/image-optimizer.ts#L186-L224), and the `getMaxAge` function is called on that. Since no max age is set, it seems like with the default loader, all optimized images served from originals in the `/public/` folder would have the same expiry as the default expiry for things in the `/public/` folder.

Do I have to use the [custom headers feature of `next.config.js`](https://nextjs.org/docs/api-reference/next.config.js/headers) to set this window to something greater than 60s?

**Reply by schlosser** [reply]
It looks like the default max age of things in the public folder is `0`? Which would explain all my images getting regenerated on so frequently... it's true that the 60s minimum would take over here, but still for static assets 60s is too short.

**Reply by synoptase** [reply]
For locally hosted assets, minimum is 60s yes. Cache will be refreshed when the 60s delay is expired. And i agree, 60s is too low even with browser caching enabled.

Let me know if you try something with the custom headers.
